core/jobs/batch_jobs/mailchimp_population_jobs.py

```diff
@@ -0,0 +1,188 @@
+        """Merging accumulators is just combining both of them into a single
+        list.
+        """
+        return list(accumulators)
+
+    def extract_output(
+        self, accumulator: List
+    ) -> List:
+        """Output is the accumulator itself."""
+        return accumulator
+
+
+class MailchimpPopulateJob(base_jobs.JobBase):
+    """One-off job for populating the mailchimp db."""
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        # Pcollection that returns the relevant config property with batch
+        # index.
+        config_property = (
+            self.pipeline
+            | 'Get all config properties' >> ndb_io.GetModels(
+                config_models.ConfigPropertyModel.get_all())
+            | 'Get the batch_index_for_mailchimp property value' >> beam.Filter(
+                lambda model: model.id == 'batch_index_for_mailchimp')
+            | 'Convert model into a dict' >> beam.Map(
+                lambda model: (model.id, model.value)))
+        batch_index_dict = beam.pvalue.AsDict(config_property)
+
+        # PCollection with all user ids that have opted in for email
+        # newsletters.
+        relevant_user_ids = (
+            self.pipeline
+            | 'Get all UserEmailPreferencesModel' >> ndb_io.GetModels(
+                user_models.UserEmailPreferencesModel.get_all().order(
+                    'created_on'))
+            | 'Filter for site_updates == true' >> beam.Filter(
+                lambda model: model.site_updates is True)
+            | 'Extract user id as tuple' >> beam.Map(
+                lambda preferences_model: (preferences_model.id, True))
+        )
+
+        # PCollection of all user ids in oppia mapped to their emails.
+        relevant_user_emails = (
+            self.pipeline
+            | 'Get all user settings models' >> ndb_io.GetModels(
+                user_models.UserSettingsModel.get_all())
+            | 'Extract user email' >> beam.Map(
+                lambda settings_model: (settings_model.id, settings_model.email)
+            )
+        )
+
+        # Merge both PCollections to map the user email to the user ids along
+        # with a bool 'True' if they have opted in for newsletter.
+        mailchimp_results = (
+            (relevant_user_emails, relevant_user_ids)
+            | 'Group by user_id' >> beam.CoGroupByKey()
+            | 'Drop user id' >> beam.Values()  # pylint: disable=no-value-for-parameter
+            | 'Filter for valid user emails' >> beam.Filter(
+                lambda email_status: (
+                    len(email_status[1]) == 1 and email_status[1][0] is True)
+            )
+            | 'Combine all valid tuples into a single list' >>
+                beam.CombineGlobally(CombineTuples())
+            | 'Send mailchimp request for current batch' >> beam.ParDo(
+                SendBatchMailchimpRequest(), batch_index_dict=batch_index_dict)
```

Comment:
Is the order guaranteed to be the same every time we run this job (so that the overall list that we're taking the index slice on is the same)?

Reply:
Yes, since I order by `created_on` in both the steps where I fetch user data from the database and no steps after that affect the data order, or even if it does (like `CoGroupByKey`), it affects it deterministically where every run would result in the same order.
The actual batch separation is done only at the last step, so it always gets the same order.

Reply:
Hmm, I'm not so sure about the order. How do you know it is deterministic? Also, why do we even need the same order every time we process the data?

Reply:
We are sending mails in batches, the index for which is input by the release coordinator in the config page. So, we need teh same order so that we can retry a batch at any point and it will affect the same users.
The only step that may mess up the order is `CoGroupByKey`, right? and even if it does, as long as the input is in the same order, wouldn't output also always be the same order?

Reply:
As many parts of the Beam code are executed on separate workers I'm not sure if we can guarantee that the order will be preserved. I'm not sure though. Some investigation in the Beam documentation might be worthwhile.